src/main/java/frc/robot/commands/defaultCommands/CoralDispenserCommand.java

```diff
@@ -2,45 +2,92 @@
 
-    public CoralDispenserCommand(CoralDispenserSubsystem coralDispenserSubsytem, Supplier<Double> xboxControllerRightTrigger, Supplier<Double> xboxControllerLeftTrigger){
-        this.coralDispenserSubsytem = coralDispenserSubsytem;
+    int fieldMin = 0;
+
+    public CoralDispenserCommand(CoralDispenserSubsystem coralDispenserSubsystem,
+            Supplier<Double> xboxControllerRightTrigger, Supplier<Double> xboxControllerLeftTrigger,
+            SwerveSubsystem swerveSubsystem) {
+        this.coralDispenserSubsystem = coralDispenserSubsystem;
         this.xboxControllerRightTrigger = xboxControllerRightTrigger;
         this.xboxControllerLeftTrigger = xboxControllerLeftTrigger;
-        addRequirements(coralDispenserSubsytem);
+        addRequirements(coralDispenserSubsystem);
+        this.swerveSubsystem = swerveSubsystem;
     }
 
     @Override
-    public void initialize() {}
+    public void initialize() {
+    }
 
     @Override
     public void execute() {
-        if(xboxControllerRightTrigger.get() > 0.5)
-            if(SensorStatus.kElevatorHeight > 8 && SensorStatus.kElevatorHeight < 10)
-                coralDispenserSubsytem.runMotors(.9, -.3);
-            else 
-                coralDispenserSubsytem.runMotors(.8, -.8);//subject to change
-        else if(xboxControllerLeftTrigger.get() > 0.5){
-                coralDispenserSubsytem.runMotors(-.3, .3);
-        }
+        if (xboxControllerRightTrigger.get() > 0.5)
+            dispenseCoral();
+        else if (xboxControllerLeftTrigger.get() > 0.5) {
+            reverseCoralDispenser();
+        } else if (isNearCoralStation() && !coralDispenserSubsystem.hasCoralAlready()) {
+            // TODO determine if we need lower or upper level to activate
+            //Controlled by button
+            dispenseCoral();
+
+        } else
+            // Do nothing with this as this stops motors
+            coralDispenserSubsystem.runMotors(0, 0);
+    }
+
+    // reverse the dispenser
+    private void reverseCoralDispenser() {
+        coralDispenserSubsystem.runMotors(-.3, .3);
+    }
+
+    private void dispenseCoral() {
+        if (SensorStatus.kElevatorHeight > ElevatorConstants.kElevatorL1Height - 1 && SensorStatus.kElevatorHeight < ElevatorConstants.kElevatorL1Height + 1)
+            coralDispenserSubsystem.runMotors(.9, -.3);
         else
-            coralDispenserSubsytem.runMotors(0, 0);
+            coralDispenserSubsystem.runMotors(.8, -.8);// subject to change
+        //sensor should flip to true
     }
 
     @Override
-    public void end(boolean interrupted) {}
+    public void end(boolean interrupted) {
+    }
 
     // Returns true when the command should end.
     @Override
     public boolean isFinished() {
         return false;
     }
-}
 
+    public boolean isNearCoralStation() {
+        Pose2d currentRobotPos = swerveSubsystem.getPose();
+        double currentY = currentRobotPos.getY();
+        double currentX = currentRobotPos.getX();
```

Comment:
Just FYI for future if you want to read a good article on var vs using the type:
https://medium.com/@victorhsr/java-to-var-or-not-to-var-that-is-the-question-10efe4989320